packages/webdriverjs/src/index.ts

```diff
@@ -223,18 +234,31 @@ export default class AxeBuilder {
       await axeRunPartial(this.driver, context, this.option)
     ];
 
+    if (frame) {
+      this.frameStack.push(frame);
+    }
+
     for (const { frameContext, frameSelector, frame } of frameContexts) {
-      let switchedFrame = false;
       try {
         assert(frame, `Expect frame of "${frameSelector}" to be defined`);
         await this.driver.switchTo().frame(frame);
-        switchedFrame = true;
-        partials.push(...(await this.runPartialRecursive(frameContext)));
+        partials.push(...(await this.runPartialRecursive(frameContext, frame)));
+        this.frameStack.pop();
         await this.driver.switchTo().parentFrame();
       } catch {
-        if (switchedFrame) {
-          await this.driver.switchTo().parentFrame();
+        /*
+          When switchTo().frame() fails using chromedriver (but not gekodriver)
+          it puts the driver into a really bad state that is impossible to
+          recover from. So we need to switch back to the main window and then
+          go back to the desired iframe context
+        */
+        const win = await this.driver.getWindowHandle();
+        await this.driver.switchTo().window(win);
+
+        for (const frameElm of this.frameStack) {
```

Comment:
Because the recursive call can manipulate `this.frameStack`, it feels a bit tricky to be confident that this is guaranteed to always restore to the frame stack as it existed without recursive calls, regardless of if/how any errors happened within the recursive call.

Maybe preferable would be to avoid using a mutable property for `this.frameStack` at all, and instead pass the `frameStack` (as `[...frameStack, frame]`) rather than the `frame` to `runPartialRecursive`? Then it would be a lot easier to be confident that this catch handler is restoring to the expected pre-state regardless of if/how any inner `runPartialRecursive` call might fail.